### PR TITLE
CE-3554: Improve infobox input fields labels in VE dialog

### PR DESCRIPTION
**Pull request by lukaszkonieczny**
https://wikia-inc.atlassian.net/browse/CE-3554

Improve infobox input fields labels in VE dialog by extracting labels from portable infobox markup

@Wikia/community-engineering @Wikia/west-wing

### CODE REVIEW

File: extensions/wikia/PortableInfobox/controllers/PortableInfoboxParserTagController.class.php [hunk 1]
```diff
@@ -134,7 +134,7 @@ protected function saveToParserOutput( \ParserOutput $parserOutput, Nodes\NodeIn
 		// (see: PortableInfoboxDataService.class.php)
 		if ( $raw ) {
 			$infoboxes = json_decode( $parserOutput->getProperty( PortableInfoboxDataService::INFOBOXES_PROPERTY_NAME ), true );
-			$infoboxes[] = [ 'data' => $raw->getRenderData(), 'sources' => $raw->getSource() ];
+			$infoboxes[] = [ 'data' => $raw->getRenderData(), 'sources' => $raw->getSource(), 'labels' => $raw->getSourceLabel() ];
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/controllers/PortableInfoboxParserTagController.class.php]: Do we still need sources in that case?

File: extensions/wikia/PortableInfobox/controllers/ApiQueryPortableInfobox.class.php [hunk 2]
```diff
@@ -30,9 +30,12 @@ protected function runOnPageSet( ApiPageSet $pageSet ) {
 				$pageSet->getResult()->addValue( [ 'query', 'pages', $id ], 'infoboxes', $inf );
 				foreach ( $parsedInfoboxes as $count => $infobox ) {
 					$s = isset( $infobox[ 'sources' ] ) ? $infobox[ 'sources' ] : [ ];
+					$l = isset( $infobox[ 'labels' ] ) ? $infobox[ 'labels' ] : [ ];
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/controllers/ApiQueryPortableInfobox.class.php]: Do we need sources here in sourceLabels case?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 3]
```diff
@@ -143,6 +147,24 @@ protected function getSourceForChildren() {
...
+		foreach ( $this->getChildNodes() as $item ) {
+			$label = $item->getSourceLabel();
+			if ( is_array( $label ) ) {
+				$result = array_merge( $result, $label );
+			} else {
+				$source = $item->getSource()[0];
+				if ( !empty( $source ) && !empty( $source ) ) {
+					$result[$source] = $label;
+				}
+			}
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: This logic could be done in `getSourceLabel`, that way it's output would be unified.

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeData.php [hunk 4]
```diff
@@ -13,4 +13,8 @@ public function getData() {
 
 		return $this->data;
 	}
+
+	public function getSourceLabel() {
+		return \Sanitizer::stripAllTags( $this->getInnerValue( $this->xmlNode->{self::LABEL_TAG_NAME} ) );
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeData.php]: How is it different from `Node::getSourceLabel`?

File: extensions/wikia/PortableInfobox/tests/nodes/NodeDataTest.php [hunk 5]
```diff
@@ -39,6 +39,32 @@ public function sourceDataProvider() {
...
+
+	public function sourceLabelDataProvider() {
+		return [
+			[ '<data source="test"></data>', [ ], '' ],
+			[ '<data source="test"><default>{{{test}}}</default></data>',
+				[ ], '' ],
+			[ '<data source="test"><label source="test">{{{test}}}</label><default>{{{test}}}</default></data>',
+				[ ], '{{{test}}}' ],
+			[ '<data source="test"><label>testLabel</label></data>', [ ], 'testLabel' ],
+			[ '<data></data>', [ ], '' ],
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeDataTest.php]: What about other elements test? `image`, `title` or `group`

**Comment by idradm** [comment]
Have you considered exchanging `getSource` with `getSourceLabel`? AFAIR ve is the only use case for sources. Additionally more tests would be really appreciated, especially for _no label_ elements.

File: extensions/wikia/PortableInfobox/tests/nodes/NodeDataTest.php [hunk 6]
```diff
@@ -39,6 +39,47 @@ public function sourceDataProvider() {
...
+				'</group>',
+				[], ['test' =>'labelInsideGroup', 'test2' =>'labelInsideGroup2'] ],
+			[ '<group>' .
+				'<title source="title"/>' .
+				'<image source="image"/>' .
+				'<data source="test"><label>labelInsideGroup</label></data>' .
+				'<data source="test2"><label>labelInsideGroup2</label></data>' .
+				'</group>',
+				[], ['test' =>'labelInsideGroup', 'test2' =>'labelInsideGroup2', 'title' => '', 'image' => ''] ]
+		];
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeDataTest.php:131]: Please add more tests, especially for edge cases like use of `<format>` or `<caption>` tags.

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 7]
```diff
@@ -31,14 +31,24 @@ public function getSource() {
...
-			return [ $source => $label ];
+		if ( count( $sources ) > 1 ) {
+			foreach ( $sources as $source ) {
+				if ( !empty( $source ) ) {
+					$sourceLabel = !empty( $label ) ? "{$label} ({$source})" : $label;
+					$sourceLabels = array_merge( $sourceLabels, [ $source => $sourceLabel ] );
+				}
+			}
+		} else {
+			if ( !empty( $sources[0] ) ) {
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: you could use `elseif` here

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 8]
```diff
@@ -31,14 +31,24 @@ public function getSource() {
...
-		$source = $this->extractSourceFromNode( $this->xmlNode )[0];
+		$sourceLabels = [];
+		$sources = $this->extractSourceFromNode( $this->xmlNode );
 		$label = \Sanitizer::stripAllTags( $this->getInnerValue( $this->xmlNode->{self::LABEL_TAG_NAME} ) );
 
-		if ( !empty( $source ) ) {
-			return [ $source => $label ];
+		if ( count( $sources ) > 1 ) {
+			foreach ( $sources as $source ) {
+				if ( !empty( $source ) ) {
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php:40]: Is this even possible?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 9]
```diff
@@ -31,14 +31,24 @@ public function getSource() {
...
+					$sourceLabel = !empty( $label ) ? "{$label} ({$source})" : $label;
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: Is this empty label case correct? Shouldn't it be
`!empty( $label ) ? "{$label} ({$source})" : $source;`
or, if you want it empty, should be explicit:
`!empty( $label ) ? "{$label} ({$source})" : '';`

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 10]
```diff
@@ -31,14 +31,24 @@ public function getSource() {
...
+					$sourceLabels = array_merge( $sourceLabels, [ $source => $sourceLabel ] );
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: Is there any reason for `array_merge` here? Why not `$sourceLabels[$source] = $sourceLablel`?
You could also make those two a one liner:
`$sourceLabels[$source] = !empty( $label ) ? "{$label} ({$source})" : $source;`

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 11]
```diff
@@ -31,14 +31,24 @@ public function getSource() {
...
+				}
+			}
+		} else {
+			if ( !empty( $sources[0] ) ) {
+				return [ $sources[0] => $label ];
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: Maybe you should add this to `sourceLabels` instead of quitting early?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 12]
```diff
@@ -38,14 +38,12 @@ public function getSourceLabel() {
 		if ( count( $sources ) > 1 ) {
 			foreach ( $sources as $source ) {
 				if ( !empty( $source ) ) {
-					$sourceLabel = !empty( $label ) ? "{$label} ({$source})" : $label;
-					$sourceLabels = array_merge( $sourceLabels, [ $source => $sourceLabel ] );
+					$sourceLabel = !empty( $label ) ? "{$label} ({$source})" : '';
+					$sourceLabels[$source] = $sourceLabel;
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: You are sure you don't want to make it a one line? `$sourceLabel` is assignment is unnecessary here.

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 13]
```diff
@@ -30,6 +30,25 @@ public function getSource() {
...
+	public function getSourceLabel() {
+		$sourceLabels = [];
+		$sources = $this->extractSourceFromNode( $this->xmlNode );
+		$label = \Sanitizer::stripAllTags( $this->getInnerValue( $this->xmlNode->{self::LABEL_TAG_NAME} ) );
+
+		if ( count( $sources ) > 1 ) {
+			foreach ( $sources as $source ) {
+				if ( !empty( $source ) ) {
+					$sourceLabel = !empty( $label ) ? "{$label} ({$source})" : '';
+					$sourceLabels[$source] = $sourceLabel;
```
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: Are you sure you don't want to make it a single line?

**Comment by idradm** [comment]
Just one remark and it's good.

Just make sure to check `PortableInfoboxDataService::getData` usages, as some other extensions might use it output, and therefore break due to format changes.

**Comment by idradm** [comment]
LGTM

**Comment by Grunny** [comment]
LGTM.
